### PR TITLE
Support Outer Join Simplify

### PR DESCRIPTION
**Pull request by XuHuaiyu**
@shenli @coocood @hanfei1991 @zimulala @tiancaiamao 
PTAL

### CODE REVIEW

**Comment by shenli** [comment]
Why close it?

**Comment by XuHuaiyu** [comment]
I'll solve a problem and reopen it today.

**Comment by XuHuaiyu** [comment]
@shenli @hanfei1991 @coocood @tiancaiamao @zimulala 
PTAL

File: plan/predicate_push_down.go [hunk 1]
```diff
@@ -77,7 +81,10 @@ func (p *NewTableDual) PredicatePushDown(predicates []expression.Expression) ([]
 
 // PredicatePushDown implements LogicalPlan PredicatePushDown interface.
 func (p *Join) PredicatePushDown(predicates []expression.Expression) (ret []expression.Expression, retPlan LogicalPlan, err error) {
-	//TODO: add null rejecter.
+	err = outerJoinSimplifier(p, predicates)
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: simplifier -> simplify

File: plan/predicate_push_down.go [hunk 2]
```diff
@@ -138,6 +145,182 @@ func (p *Join) PredicatePushDown(predicates []expression.Expression) (ret []expr
...
+			err := outerJoinSimplifier(x, fullConditions)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+	return nil
+}
+
+// isNullRejected check whether a condition is null-rejected
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: Add preriod for every comment.

File: plan/predicate_push_down.go [hunk 3]
```diff
@@ -138,6 +145,182 @@ func (p *Join) PredicatePushDown(predicates []expression.Expression) (ret []expr
...
+				isOk, err := isNullRejected(schema, x)
+				if err != nil || !isOk {
+					return false, errors.Trace(err)
+				}
+			}
+		}
+		return true, nil
+	}
+	// ignore control functions
+	if _, ok := evaluator.CntrlFuncs[scalarFunc.FuncName.L]; ok {
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: why ignore them ?

File: plan/predicate_push_down.go [hunk 4]
```diff
@@ -138,6 +144,176 @@ func (p *Join) PredicatePushDown(predicates []expression.Expression) (ret []expr
...
+		outerTable = child1
+	} else if p.JoinType == RightOuterJoin {
+		innerTable = child1
+		outerTable = child2
+	}
+	switch innerPlan := innerTable.(type) {
+	case *DataSource:
+		canBeSimplified := false
+		for _, expr := range predicates {
+			if canBeSimplified {
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: why not move this logic to line 205?

File: plan/predicate_push_down.go [hunk 5]
```diff
@@ -138,6 +144,176 @@ func (p *Join) PredicatePushDown(predicates []expression.Expression) (ret []expr
...
+		return nil
+	} else if p.JoinType == LeftOuterJoin {
+		innerTable = child2
+		outerTable = child1
+	} else if p.JoinType == RightOuterJoin {
+		innerTable = child1
+		outerTable = child2
+	}
+	switch innerPlan := innerTable.(type) {
+	case *DataSource:
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: You needn't judge the type of join's children.

File: plan/predicate_push_down.go [hunk 6]
```diff
@@ -138,6 +144,176 @@ func (p *Join) PredicatePushDown(predicates []expression.Expression) (ret []expr
 	return
 }
 
+// outerJoinSimplifier simplify outer join.
```
Comment (zimulala) [inline plan/predicate_push_down.go]: simplify -> simplifies

File: plan/predicate_push_down.go [hunk 7]
```diff
@@ -138,6 +144,176 @@ func (p *Join) PredicatePushDown(predicates []expression.Expression) (ret []expr
...
+	}
+	return nil
+}
+
+// isNullRejected check whether a condition is null-rejected
+// A condition would be null-rejected in one of following cases:
+// If it is a predicate containing a reference to an inner table that evaluates to UNKNOWN or FALSE when one of its arguments is NULL.
+// If it is a conjunction containing a null-rejected condition as a conjunct.
+// If it is a disjunction of null-rejected conditions.
+func isNullRejected(schema expression.Schema, scalarFunc *expression.ScalarFunction) (bool, error) {
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: you should combine isNullRejected and calculateResultOfScalarFunc to one function, return nil that stands for unknown result.

File: plan/predicate_push_down.go [hunk 8]
```diff
@@ -138,6 +144,176 @@ func (p *Join) PredicatePushDown(predicates []expression.Expression) (ret []expr
...
+		for _, expr := range predicates {
+			if canBeSimplified {
+				break
+			}
+			switch x := expr.(type) {
+			case *expression.Constant:
+				if x.Value.IsNull() {
+					canBeSimplified = true
+				} else if isTrue, err := x.Value.ToBool(); err != nil || isTrue == 0 {
+					if err != nil {
```
Comment (zimulala) [inline plan/predicate_push_down.go]: Five layers of nested isn't elegant.

File: plan/predicate_push_down.go [hunk 9]
```diff
@@ -138,6 +143,119 @@ func (p *Join) PredicatePushDown(predicates []expression.Expression) (ret []expr
...
+// calculateResultOfExpression set inner table columns in a expression as null and calculate the finally result of the scalar function.
+func calculateResultOfExpression(schema expression.Schema, expr expression.Expression) (expression.Expression, error) {
+	switch x := expr.(type) {
+	case *expression.ScalarFunction:
+		var err error
+		args := make([]expression.Expression, len(x.Args))
+		for i, arg := range x.Args {
+			args[i], err = calculateResultOfExpression(schema, arg)
+		}
+		if err != nil {
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: when the error is not nil?

**Comment by hanfei1991** [comment]
LGTM

File: plan/predicate_push_down.go [hunk 10]
```diff
@@ -138,6 +143,119 @@ func (p *Join) PredicatePushDown(predicates []expression.Expression) (ret []expr
...
+		return x.DeepCopy(), nil
+	}
+}
+
+// concatOnAndWhereConds concatenate ON conditions with WHERE conditions.
+func concatOnAndWhereConds(join *Join, predicates []expression.Expression) []expression.Expression {
+	equalConds, leftConds, rightConds, otherConds := join.EqualConditions, join.LeftConditions, join.RightConditions, join.OtherConditions
+	ans := make([]expression.Expression, 0, len(equalConds)+len(leftConds)+len(rightConds)+len(predicates))
+	for _, v := range equalConds {
+		ans = append(ans, v)
```
Comment (shenli) [inline plan/predicate_push_down.go]: Why not use ans = append(ans, equalConds...)?

**Comment by shenli** [comment]
LGTM

**Pull request merged**